Lib/test/test_compileall.py

```diff
@@ -825,6 +1054,223 @@ def test_ignore_symlink_destination(self):
         self.assertTrue(os.path.isfile(allowed_bc))
         self.assertFalse(os.path.isfile(prohibited_bc))
 
+    def test_hardlink_deduplication_bad_args(self):
+        # Bad arguments combination, hardlink deduplication make sense
+        # only for more than one optimization level
+        self.assertRunNotOK(self.directory, "-o 1", "--hardlink_dupes")
+
+    def test_hardlink_deduplication_same_bytecode_all_opt(self):
+        # 'a = 0' produces the same bytecode for all optimization levels
+        path = os.path.join(self.directory, "test", "same_all")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode",
+                                                  "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt1 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=1)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=2)
+
+        self.assertRunOK(path, "-q", "-o 0", "-o 1", "-o 2",
+                         "--hardlink-dupes")
+
+        # All three files should have the same inode (hardlinks)
+        self.assertEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
+        self.assertEqual(os.stat(pyc_opt1).st_ino, os.stat(pyc_opt2).st_ino)
+
+        for pyc_file in {pyc_opt0, pyc_opt1, pyc_opt2}:
+            os.unlink(pyc_file)
+
+        self.assertRunOK(path, "-q", "-o 0", "-o 1", "-o 2")
+
+        # Deduplication disabled, all pyc files should have different inodes
+        self.assertNotEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
+        self.assertNotEqual(os.stat(pyc_opt1).st_ino, os.stat(pyc_opt2).st_ino)
+
+    def test_hardlink_deduplication_same_bytecode_some_opt(self):
+        # 'a = 0' produces the same bytecode for all optimization levels
+        # only two levels of optimization [0, 1] tested
+        path = os.path.join(self.directory, "test", "same_some")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode",
+                                                  "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=2)
+
+        self.assertRunOK(path, "-q", "-o 0", "-o 2", "--hardlink-dupes")
+
+        # Both files should have the same inode (hardlink)
+        self.assertEqual(os.stat(pyc_opt0).st_ino,  os.stat(pyc_opt2).st_ino)
+
+        for pyc_file in {pyc_opt0, pyc_opt2}:
+            os.unlink(pyc_file)
+
+        self.assertRunOK(path, "-q", "-o 0", "-o 2")
+
+        # Deduplication disabled, both pyc files should have different inodes
+        self.assertNotEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt2).st_ino)
```

Comment:
Code looks copied from test_hardlink_deduplication_same_bytecode_all_opt(). Try to factorize duplicated code.